### ISSUE TITLE
Development Mode with multiple maven modules not working

### ISSUE THREAD

**Issue by uberto** [quarkusio/quarkus#10677]
**Describe the bug**
I'm not sure if this is a genuine issue or a known limitation.
I also checked in the doc but I couldn't find any reference to it.

I've tried to re-create our production project module organisation in a mini project on github:
https://github.com/uberto/h4kcluster-quarkus

Now we have different modules for domain, adapter, ui and the actual applications.

We don't have any source on the parent project.

Maven build is fine and when I start quarkus with dev mode specifying "launcher" (the project with the servlet as quarkus entry point) it starts correctly:
./mvnw -pl launcher quarkus:dev

It will also detect and hot-reload the changes I do to the launcher project, but it is not detecting if I change something else. I would need to stop, "mvn install" and restart to detect those.

Am I missing something?

Would it be better using Gradle? since it has direct project dependencies and doesn't require an install?

**Expected behavior**
Changing a submodule, for example "domain" should be reflected immediately when running quarkus in dev mode

**Actual behavior**
I need to stop dev mode and mvn install to see the change

**To Reproduce**
1. clone https://github.com/uberto/h4kcluster-quarkus
2. ./mvnw -pl launcher quarkus:dev
3. try to change something in domain package and see the it's reflected in the running app

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Linux pop-os 5.4.0-7634-generic #38~1592497129~20.04~9a1ea2e-Ubuntu SMP Fri Jun 19 22:43:37 UTC  x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: 
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.20.0, JRE 14 Linux amd64-64-Bit Compressed References 20200416_44 (JIT enabled, AOT enabled)
OpenJ9   - 05fa2d361
OMR      - d4365f371
JCL      - 5757187cae based on jdk-14.0.1+7)

 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 
1.6.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Maven:  Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/ubertobarbini/.m2/wrapper/dists/apache-maven-3.6.3-bin/1iopthnavndlasol9gbrbg6bf2/apache-maven-3.6.3
Java version: 14.0.1, vendor: AdoptOpenJDK, runtime: /home/ubertobarbini/.sdkman/candidates/java/14.0.1.j9-adpt
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.0-7634-generic", arch: "amd64", family: "unix"

**Comment by aloubyansky** [comment]
And w/o `-pl`?

**Comment by ubertobarbini** [comment]
> And w/o `-pl`?

then quarkus:dev would fail with "null pointer execption"
(there are no sources in the main project)

**Comment by famod** [comment]
That NPE is probably the one from #10607?

What happens when using `-f launcher` or `cd launcher`?

**Comment by aloubyansky** [comment]
Right, `-pl` is excluding projects from the Maven reactor, iirc.

**Comment by ubertobarbini** [comment]
with -f I got this error:

Could not resolve dependencies for project com.ubertob.h4kcluster:launcher:jar:1.0.0-SNAPSHOT: Could not find artifact com.ubertob.h4kcluster:sumnumbers-app:jar:1.0.0-SNAPSHOT in jcenter (https://jcenter.bintray.com/) 

(it cannot solve launcher dependencies on other subprojects: sumnumbers-app is not on jcenter)

I've same result if I try to launch quarkus:dev from the launcher folder

fyi this is the pom of launcher project:
```

<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
<parent>
<artifactId>h4kcluster-quarkus</artifactId>
<groupId>com.ubertob.h4kcluster</groupId>
<version>1.0.0-SNAPSHOT</version>
</parent>
<modelVersion>4.0.0</modelVersion>
<artifactId>launcher</artifactId>
<dependencies>
<dependency>
<groupId>com.ubertob.h4kcluster</groupId>
<artifactId>sumnumbers-app</artifactId>
<version>1.0.0-SNAPSHOT</version>
</dependency>
</dependencies>
</project>
```

**Comment by aloubyansky** [comment]
Is `sumnumbers-app` in the same workspace? If so has it been compiled? Or packaged? Is it installed in the local repo?
Normally, `mvn clean compile quarkus:dev` should work. If it can't find a local dependencies it means it hasn't been built (compiled or packaged) or installed into the local repo.

**Comment by ubertobarbini** [comment]
`sumnumbers-app` is a module in the same project.
If I do a mvn install first then 
` mvn quarkus:dev -f launcher`

it starts the app but it still not detecting any changes I'm doing in `domain` or in `sumnumbers-app`

In gradle I can specify dependencies between "siblings" using project dependecies but there is nothing like that in maven, afaik.

On the other side since once it runs everything should be in the classpath, it shouldn't matter how it went there... so I don't know.

Btw correction:
if I run
` mvn compile quarkus:dev`
(without -f) it starts but it doesn't show my servler and instead it's showing an error web page:
<html><head><title>Error</title></head><body>Not Found</body></html>

**Comment by ubertobarbini** [comment]
```
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ launcher ---
[INFO] Building jar: /home/ubertospringer/svi/h4kcluster-quarkus/launcher/target/launcher-1.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- quarkus-maven-plugin:1.6.0.Final:build (default) @ launcher ---
[INFO] [org.jboss.threads] JBoss Threads version 3.1.1.Final
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building thin jar: /home/ubertospringer/svi/h4kcluster-quarkus/launcher/target/launcher-1.0.0-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 898ms
[INFO] 
[INFO] --- maven-install-plugin:2.4:install (default-install) @ launcher ---
[INFO] Installing /home/ubertospringer/svi/h4kcluster-quarkus/launcher/target/launcher-1.0.0-SNAPSHOT.jar to /home/ubertospringer/.m2/repository/com/ubertob/h4kcluster/launcher/1.0.0-SNAPSHOT/launcher-1.0.0-SNAPSHOT.jar
[INFO] Installing /home/ubertospringer/svi/h4kcluster-quarkus/launcher/pom.xml to /home/ubertospringer/.m2/repository/com/ubertob/h4kcluster/launcher/1.0.0-SNAPSHOT/launcher-1.0.0-SNAPSHOT.pom
[INFO] 
[INFO] --- quarkus-maven-plugin:1.6.0.Final:dev (default-cli) @ launcher ---
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
Listening for transport dt_socket at address: 5005
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-07-13 17:18:25,174 INFO  [io.quarkus] (Quarkus Main Thread) launcher 1.0.0-SNAPSHOT on JVM (powered by Quarkus 1.6.0.Final) started in 0.750s. Listening on: http://0.0.0.0:8080
2020-07-13 17:18:25,185 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-07-13 17:18:25,185 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, kotlin, servlet]
```

this is the console output at startup

if I try to change something in `domain` or `sumnumber-app` there is not notification...

but if I touch the `launcher` module I can see:
```
2020-07-13 17:19:56,924 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-2) Changed source files detected, recompiling [/home/ubertospringer/svi/h4kcluster-quarkus/launcher/src/main/kotlin/com/ubertob/h4kcluster/EgressServlet.kt]
2020-07-13 17:19:58,208 INFO  [io.quarkus] (Quarkus Main Thread) launcher stopped in 0.040s
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2020-07-13 17:19:58,384 INFO  [io.quarkus] (Quarkus Main Thread) launcher 1.0.0-SNAPSHOT on JVM (powered by Quarkus 1.6.0.Final) started in 0.172s. Listening on: http://0.0.0.0:8080
2020-07-13 17:19:58,385 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2020-07-13 17:19:58,385 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, kotlin, servlet]
2020-07-13 17:19:58,385 INFO  [io.qua.dep.dev.RuntimeUpdatesProcessor] (vert.x-worker-thread-2) Hot replace total time: 1.462s 

```

So, I wonder if it's a problem with source location, since it's a Kotlin project...

**Comment by ubertobarbini** [comment]
Is there a way to know the list of the folder where quarkus is listening for changes?

**Comment by uberto** [comment]
Yay! the issue was on the different source:

adding 
```
    <build>
        <sourceDirectory>src/main/kotlin</sourceDirectory>
        <testSourceDirectory>src/test/kotlin</testSourceDirectory>
    </build>
```
to the child pom fixed the hot replace issue.

I think it's still a minor bug since it shouldn't be needed (maven compile perfectly wihtout it)

The other not very cool issue is that the time for hot replace is now in a completely different ballpark than with a single module (same quantity of code) :(
```
Hot replace total time: 2.892s 
```

**Comment by gsmet** [comment]
/cc @evanchooly for the Kotlin issue.

> Hot replace total time: 2.892s

Could you try to get more info via https://github.com/quarkusio/quarkus/blob/master/TROUBLESHOOTING.md#profiling-application-dev-mode-with-async-profiler ?

**Comment by uberto** [comment]
I realized I was using J9 for these tests, going back to OpenJdk 14 gave me better timings.

the first hot-replace is still quite slow (1.5 s.) but next call is 0.5s and then progressively faster until it stabilizes on 320 ms.

I generated the svg file from the profile agent (btw nice improvement on old way to have flamegraphs).
Is there anything specific I should look at it?



Btw about the launching:
`./mvnw  quarkus:dev`

gives me:
```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.6.0.Final:dev (default-cli) on project h4kcluster-quarkus: Execution default-cli of goal io.quarkus:quarkus-maven-plugin:1.6.0.Final:dev failed.: NullPointerException 
```

`./mvnw compile quarkus:dev`
doesn't log errors but the application only show an not found page:
```
<html><head><title>Error</title></head><body>Not Found</body></html>
```

Instead
`./mvnw -f launcher quarkus:dev`
works perfectly, without the compile.
It's not a big issue but it's a bit confusing
